File: app/src/main/java/com/capstone/project_niyakneyak/ui/login/LoginViewModel.java
package com.capstone.project_niyakneyak.ui.login;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import android.util.Patterns;

import com.capstone.project_niyakneyak.data.LoginRepository;
import com.capstone.project_niyakneyak.data.Result;
import com.capstone.project_niyakneyak.data.model.LoggedInUser;
import com.capstone.project_niyakneyak.R;

public class LoginViewModel extends ViewModel {

    private MutableLiveData<LoginFormState> loginFormState = new MutableLiveData<>();
    private MutableLiveData<LoginResult> loginResult = new MutableLiveData<>();
    private LoginRepository loginRepository;

    LoginViewModel(LoginRepository loginRepository) {
        this.loginRepository = loginRepository;
    }

    LiveData<LoginFormState> getLoginFormState() {
        return loginFormState;
    }

    LiveData<LoginResult> getLoginResult() {
        return loginResult;
    }

    public void login(String username, String password) {
        // can be launched in a separate asynchronous job
        Result<LoggedInUser> result = loginRepository.login(username, password);

        if (result instanceof Result.Success) {
            LoggedInUser data = ((Result.Success<LoggedInUser>) result).getData();
            loginResult.setValue(new LoginResult(new LoggedInUserView(data.getDisplayName())));
        } else {
            loginResult.setValue(new LoginResult(R.string.login_failed));
        }
    }

    public void loginDataChanged(String username, String password) {
        if (!isUserNameValid(username)) {
            loginFormState.setValue(new LoginFormState(R.string.invalid_username, null));
        } else if (!isPasswordValid(password)) {
            int errorCode = (password.length() <= 5) ? R.string.invalid_password_less : R.string.invalid_password_more;
            loginFormState.setValue(new LoginFormState(null, errorCode));
        } else {
            loginFormState.setValue(new LoginFormState(true));
        }
    }

    // A placeholder username validation check
    private boolean isUserNameValid(String username) {
        if (username == null) {
            return false;
        }
        if (username.contains("@")) {
            return Patterns.EMAIL_ADDRESS.matcher(username).matches();
        } else {
            return !username.trim().isEmpty();
        }
    }

    // A placeholder password validation check
    private boolean isPasswordValid(String password) {
        return password != null && (password.trim().length() > 5 && password.trim().length() <= 20);
    }
}

File: app/src/main/java/com/capstone/project_niyakneyak/login/viewmodel/LoginViewModel.kt
package com.capstone.project_niyakneyak.login.viewmodel

import android.util.Patterns
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.capstone.project_niyakneyak.R
import com.capstone.project_niyakneyak.login.etc.LoginFormState
import com.capstone.project_niyakneyak.login.etc.LoginResult

class LoginViewModel : ViewModel() {
    val loginFormState = MutableLiveData<LoginFormState?>()
    val loginResult = MutableLiveData<LoginResult?>()

    fun loginDataChanged(username: String?, password: String) {
        if (!isUserNameValid(username) && !isPasswordValid(password)) {
            val errorCode =
                if (password.length <= 5) R.string.invalid_password_less else R.string.invalid_password_more
            loginFormState.setValue(LoginFormState(R.string.invalid_username, errorCode))
        } else if (!isUserNameValid(username)) {
            loginFormState.setValue(LoginFormState(R.string.invalid_username, null))
        } else if (!isPasswordValid(password)) {
            val errorCode =
                if (password.length <= 5) R.string.invalid_password_less else R.string.invalid_password_more
            loginFormState.setValue(LoginFormState(null, errorCode))
        } else {
            loginFormState.setValue(LoginFormState(true))
        }
    }

    // A placeholder username validation check
    private fun isUserNameValid(username: String?): Boolean {
        if (username == null) {
            return false
        }
        return if (username.contains("@")) {
            Patterns.EMAIL_ADDRESS.matcher(username).matches()
        } else {
            !username.trim { it <= ' ' }.isEmpty() && username.trim { it <= ' ' }.length <= 30
        }
    }

    // A placeholder password validation check
    private fun isPasswordValid(password: String?): Boolean {
        return password != null && password.trim { it <= ' ' }.length > 5 && password.trim { it <= ' ' }.length <= 20
    }
}